grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,280 @@
+    }
+
+    private static HealthCheckResponse getHealthCheckResponse(ServingStatus status) {
+        if (status == ServingStatus.SERVING) {
+            return SERVING_RESPONSE;
+        } else if (status == ServingStatus.NOT_SERVING) {
+            return NOT_SERVING_RESPONSE;
+        } else if (status == ServingStatus.SERVICE_UNKNOWN) {
+            return SERVICE_UNKNOWN_RESPONSE;
+        } else {
+            throw new IllegalArgumentException("Invalid status:" + status);
+        }
+    }
+
+    private final SettableHealthChecker serverHealth;
+    private final Set<ListenableHealthChecker> healthCheckers;
+    private final Map<String, ListenableHealthChecker> grpcServiceHealthCheckers;
+    private final NonBlockingIdentityHashMap<StreamObserver<HealthCheckResponse>, Entry<String, ServingStatus>>
+            watchers = new NonBlockingIdentityHashMap<>();
+    @Nullable
+    private Server server;
+    private boolean updating;
+
+    GrpcHealthCheckService(
+            Set<ListenableHealthChecker> healthCheckers,
+            Map<String, ListenableHealthChecker> grpcServiceHealthCheckers,
+            List<HealthCheckUpdateListener> updateListeners
+    ) {
+        serverHealth = new SettableHealthChecker(false);
+        if (!updateListeners.isEmpty()) {
+            addServerHealthUpdateListener(updateListeners);
+        }
+        this.healthCheckers = ImmutableSet.<ListenableHealthChecker>builder()
+                                          .add(serverHealth)
+                                          .addAll(healthCheckers)
+                                          .build();
+        this.grpcServiceHealthCheckers = grpcServiceHealthCheckers;
+        final HealthCheckUpdateListener healthCheckUpdateListener = provideInternalHealthUpdateListener();
+        setInternalHealthUpdateListener(healthCheckUpdateListener, this.healthCheckers);
+        setInternalHealthUpdateListener(healthCheckUpdateListener, this.grpcServiceHealthCheckers.values());
+    }
+
+    @Override
+    public void check(HealthCheckRequest request, StreamObserver<HealthCheckResponse> responseObserver) {
+        final String service = request.getService();
+        final ServingStatus status = checkServingStatus(service);
+        if (status == ServingStatus.SERVICE_UNKNOWN) {
+            responseObserver.onError(Status.NOT_FOUND
+                                             .withDescription(String.format(
+                                                     "The service name(%s) is not registered in this service",
+                                                     service))
+                                             .asRuntimeException());
+            return;
+        }
+        responseObserver.onNext(getHealthCheckResponse(status));
+        responseObserver.onCompleted();
+    }
+
+    @Override
+    public void watch(HealthCheckRequest request, StreamObserver<HealthCheckResponse> responseObserver) {
```

Comment:
One more thing we have to check is that the request shouldn't be aborted after 15 seconds request timeout.
Could you check if it's working as expected by adding a test?
If it's not, I guess we have to disable the request timeout or increase the timeout to a sensible default.

Reply:
Good point. I missed that.

> disable the request timeout 

+1. It is doable by calling `ServiceRequestContext.current().clearRequestTimeout()`. 
1) Upstream implementation does not have a deadline for `watch`es.
2) Most applications would expose a health check service to secure zones and internal gateways. It seems to me that it is safe not to have a timeout for it.

Reply:
> Could you check if it's working as expected by adding a test?